### PR TITLE
Mark "React" symbol as referenced

### LINKED ISSUE #10312: --noUnusedLocals doesn't consider React to be "used" when using JSX files
**TypeScript Version:**   2.1.0-dev.20160723 

... because it complains that `React` is declared but never used. It doesn't understand JSX transformations apparently.

### PR DESCRIPTION
**Pull request by sheetalkamat**
It won't not be marked as ```isReferenced``` in ```resolveName``` when there is no ```diagnosticMessage``` when resolving it.
Fixes #10312

### CODE REVIEW

File: tests/cases/compiler/unusedImports14.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
+//@module: commonjs
+//@jsx: react
```
Comment (mhegazy) [inline tests/cases/compiler/unusedImports14.ts]: can you add a test for `reactNamesapce` and `noUnusedLocals`

File: tests/cases/compiler/unusedImports14.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
```
Comment (mhegazy) [inline tests/cases/compiler/unusedImports14.ts]: you do not need the unused params here.

**Comment by bartq** [comment]
what about custom JSX backend using `reactNamespace` param in `tsconfig.json`? Is it handled here too?

**Comment by mhegazy** [comment]
> what about custom JSX backend using reactNamespace param in tsconfig.json? Is it handled here too?

yup. see the test in https://github.com/Microsoft/TypeScript/blob/67873ca20b334c0b21e2a70440ab431a8b4bc41b/tests/cases/compiler/unusedImports15.ts

**Comment by fspillner** [comment]
It seems, that it doesn't work with new jsxFactory param in tsconfig.json. Can somebody confirm this? Should I open new issue?

**Comment by mhegazy** [comment]
> Should I open new issue?

yes please.

**Comment by fspillner** [comment]
@mhegazy here you go: https://github.com/Microsoft/TypeScript/issues/14109

**Pull request merged**